src/Interpreters/InterpreterSelectQueryAnalyzer.cpp

```diff
@@ -74,60 +75,28 @@ ContextMutablePtr buildContext(const ContextPtr & context, const SelectQueryOpti
 
 void replaceStorageInQueryTree(QueryTreeNodePtr & query_tree, const ContextPtr & context, const StoragePtr & storage)
 {
-    auto query_to_replace_table_expression = query_tree;
-    QueryTreeNodePtr table_expression_to_replace;
+    auto nodes = extractTrueTableExpressions(query_tree);
+    IQueryTreeNode::ReplacementMap replacement_map;
 
-    while (!table_expression_to_replace)
+    for (auto & node : nodes)
     {
-        if (auto * union_node = query_to_replace_table_expression->as<UnionNode>())
-            query_to_replace_table_expression = union_node->getQueries().getNodes().at(0);
-
-        auto & query_to_replace_table_expression_typed = query_to_replace_table_expression->as<QueryNode &>();
-        auto left_table_expression = extractLeftTableExpression(query_to_replace_table_expression_typed.getJoinTree());
-        auto left_table_expression_node_type = left_table_expression->getNodeType();
-
-        switch (left_table_expression_node_type)
-        {
-            case QueryTreeNodeType::QUERY:
-            case QueryTreeNodeType::UNION:
-            {
-                query_to_replace_table_expression = std::move(left_table_expression);
-                break;
-            }
-            case QueryTreeNodeType::TABLE:
-            case QueryTreeNodeType::TABLE_FUNCTION:
-            case QueryTreeNodeType::IDENTIFIER:
-            {
-                table_expression_to_replace = std::move(left_table_expression);
-                break;
-            }
-            default:
-            {
-                throw Exception(ErrorCodes::UNSUPPORTED_METHOD,
-                    "Expected table, table function or identifier node to replace with storage. Actual {}",
-                    left_table_expression->formatASTForErrorMessage());
-            }
-        }
-    }
+        auto & table_node = node->as<TableNode &>();
 
-    /// Don't replace storage if table name differs
-    if (auto * table_node = table_expression_to_replace->as<TableNode>(); table_node && table_node->getStorageID().getFullNameNotQuoted() != storage->getStorageID().getFullNameNotQuoted())
-        return;
+        /// Don't replace storage if table name differs
+        if (table_node.getStorageID().getFullNameNotQuoted() != storage->getStorageID().getFullNameNotQuoted())
```

Comment:
Wouldn't it be better to compare `uuid`s? Though, it's difficult to imagine a case when comparing full names is a problem.